### PR TITLE
Grey card undetermine state without wifi (DEV, EXPOSUREAPP-3892)

### PR DESCRIPTION
**Pull request by BMItr**
Quick fix for grey card state (1.8.x RC)

### CODE REVIEW

**Comment by BMItter** [comment]
issue is under investigation. (DRAFT)

**Review by kolyaopahle — DISMISSED** [review]
Worked on my device 👍

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/RiskLevelRepository.kt [hunk 1]
```diff
@@ -7,7 +7,8 @@ import kotlinx.coroutines.flow.MutableStateFlow
 
 object RiskLevelRepository {
 
-    private val internalRisklevelScore = MutableStateFlow(getLastSuccessfullyCalculatedScore().raw)
+    private val internalRisklevelScore = MutableStateFlow(if (getLastSuccessfullyCalculatedScore() !=
```
Comment (ralfgehrer) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/RiskLevelRepository.kt:10]: I'd recommend moving this into a function call.

Also, getLastSuccessfullyCalculatedScore() is called twice, please "cache" it in a val.

Most importantly, I don't see how why this value has to be overwritten in this particular Flow initializer. getLastSuccessfullyCalculatedScore() solely accesses the LocalData. UNDETERMINED  is a default value if no data have been found/stored or the risk value integer value cannot be mapped to a constant. IMHO this is not addressing the issue in the right place. How does this relate to the no-wifi problem?

Lastly, please ensure it does not break the fixed behavior of EXPOSUREAPP-3723.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/RiskLevelRepository.kt [hunk 2]
```diff
@@ -7,7 +7,8 @@ import kotlinx.coroutines.flow.MutableStateFlow
...
+        RiskLevel.UNDETERMINED) getLastSuccessfullyCalculatedScore().raw else RiskLevel.UNKNOWN_RISK_INITIAL.raw)
```
Comment (ralfgehrer) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/RiskLevelRepository.kt:11]: I'd recommend moving this into a function call.

Also, getLastSuccessfullyCalculatedScore() is called twice, please "cache" it in a val.

Most importantly, I don't see how why this value has to be overwritten in this particular Flow initializer. getLastSuccessfullyCalculatedScore() solely accesses the LocalData. UNDETERMINED is a default value if no data have been found/stored or the risk value integer value cannot be mapped to a constant. IMHO this is not addressing the issue in the right place. How does this relate to the no-wifi problem?

Lastly, please ensure it does not break the fixed behavior of EXPOSUREAPP-3723.

**Review by ralfgehrer — CHANGES_REQUESTED** [review]

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/RiskLevelRepository.kt (hunk 2 above)
Comment (BMItter) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/RiskLevelRepository.kt:11]: there were just some minutes left on Friday for trying to fix this. I'm going to optimize the code. Will also check Exposureapp-3723 for collisions. If it does, further investigations by the original author are required.

Comment (Oliver-Zimmerman) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/RiskLevelRepository.kt:11]: I think it should be if(getLastSuccessfullyCalculatedScore().raw != RiskLevel.UNDETERMINED) getLastSuccesfullCalculatesScore.raw() else RiskLevelConstants.UNKNOWN_RISK_INITIAL

This more closely aligns with the original and the changes made here:
https://github.com/corona-warn-app/cwa-app-android/pull/1597/files

I can test when I get a chance

Comment (Oliver-Zimmerman) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/RiskLevelRepository.kt:11]: Although I think they effectively do the same thing.

**Review by Oliver-Zimmerman — DISMISSED** [review]
Works for me and seems to resolve EXPOSUREAPP-3892 as well. 👍

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/RiskLevelRepository.kt (hunk 2 above)
Comment (BMItter) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/RiskLevelRepository.kt]: same effect.

**Review by kolyaopahle — DISMISSED** [review]

**Review by Oliver-Zimmerman — DISMISSED** [review]

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/RiskLevelRepository.kt [hunk 3]
```diff
@@ -7,14 +7,28 @@ import kotlinx.coroutines.flow.MutableStateFlow
...
     private val internalRiskLevelScoreLastSuccessfulCalculated =
         MutableStateFlow(LocalData.lastSuccessfullyCalculatedRiskLevel().raw)
     val riskLevelScoreLastSuccessfulCalculated: Flow<Int> =
         internalRiskLevelScoreLastSuccessfulCalculated
 
+    /**
+     * Set the internal risk level score [internalRisklevelScore]
+     */
+    private fun setInternalRiskLevelScore(): MutableStateFlow<Int> {
+        val lastSuccessfullyCalculatedScore = getLastCalculatedScore().raw
```
Comment (ralfgehrer) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/RiskLevelRepository.kt:22]: Thanks! I still don't get though, why we do this here in the Repo and not at the source in  LocalData.lastCalculatedRiskLevel().
Currently, we have overwriting behavior in multiple places.

**Review by chris-cwa — DISMISSED** [review]

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/RiskLevelRepository.kt (hunk 3 above)
Comment (Oliver-Zimmerman) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/RiskLevelRepository.kt]: No, I think you're correct and I have adjusted the code to match this.

**Comment by BMItter** [comment]
device Tests looking good so far (with & without connection). We still need to check for sideeffects...

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/RiskLevelRepository.kt (hunk 3 above)
Comment (BMItter) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/RiskLevelRepository.kt]: @ralfgehrer @InterestedParties 
what do you think about the UNDETERMINED state now?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/RiskLevelRepository.kt [hunk 4]
```diff
@@ -14,7 +14,6 @@ object RiskLevelRepository {
         MutableStateFlow(LocalData.lastSuccessfullyCalculatedRiskLevel().raw)
```
Comment (ralfgehrer) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/RiskLevelRepository.kt:14]: Here we access "LocalData.lastSuccessfullyCalculatedRiskLevel().raw" directly whereas we use "getLastSuccessfullyCalculatedScore().raw" in line 10 
We should align this.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/RiskLevelRepository.kt (hunk 3 above)
Comment (ralfgehrer) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/RiskLevelRepository.kt]: @BMItter not quite sure I understand your question.

This is what it says in our code. Hence, initializing sth with that value is probably not the best idea.
`  
    // mapped to no UI state
    // this should never happen
    UNDETERMINED(RiskLevelConstants.UNDETERMINED);
`

**Comment by harambasicluka** [comment]
In https://github.com/corona-warn-app/cwa-app-android/pull/1699 the UNKOWN_RISKS ARE removed, so do we even need this PR?

@BMItter @SamuraiKek @ralfgehrer @d4rken

**Comment by d4rken** [comment]
Maybe wait with this until #1705 is merged. The whole `RiskLevelRepository` will be removed. There will be single source of truth for all risk level things.

**Comment by d4rken** [comment]
After #1705 and #1699 is merged, we should reevaluate this bug again. Most of the bugs of this type were caused by weird states due to some values being stale, most of that will be fixed with these 2 PRs.

**Comment by harambasicluka** [comment]
Ticket was set to "obsolete", will close this PR.

**Pull request closed without merging**